### ISSUE TITLE
[Tracking] Module mocking in storybook

### ISSUE THREAD

**Issue by kasperpeulen** [storybookjs/storybook#26637]
Milestone 1 here is to setup a demo where we can validate all the APIs we are going to make in this project. 
In parallel we will work on milestone 2-6 which we ship with canary releases that we then can test and validate in the demo.
Milestone 2-3 are the must haves. And milestone 4-6 are nice to haves for this project.
Once the APIs are validated and finalized we can work on the docs and make a blog post of the demo.

```[tasklist]
### 1 Public demo showing how RSC's touching DB and cookies can be tested in isolation in Storybook (Yann)
- [x] Find a nextjs app router example demo that we can expand on, using DB for data access, and where at least some endpoints needs authentication (using cookies)
- [x] Change all absolute imports to standard-based package.json imports
- [ ] https://github.com/storybookjs/storybook/issues/26725
- [x] Experiment with a good API for mocking out `next/headers` so that cookies can be easily mocked in storybook, and make a good example
- [ ] https://github.com/storybookjs/storybook/issues/26808
- [ ] https://github.com/storybookjs/storybook/pull/26886
- [x] Middleware used in the demo seems a bit outdated pattern with server actions? Remove that for best practices
- [x] Make mock types: query, action and only show action by default
- [x] Rerun async components on redirect using RedirectBoundary of next
- [ ] Use the new react APIs useFormState, useFormStatus for server-side form validation in the demo (validating the notes for example) (Kasper)
- [ ] Make the demo actual secure by checking if the username matches the encrypted username (Kasper)
- [ ] The next.js RedirectBoundary causes flashing UI when used client side. Possible use the implementation of the PR: https://github.com/vercel/next.js/pull/49439 Or wait for an upstream fix? Nice to have (Kasper)
- [x] Make stories for edit routes. Possibly move tests of note-ui stories to page stories instead, discuss best practices here. (Jeppe)
- [x] Add the test runner and track coverage? Coverage number to share in the blog post? (Yann)
- [x] Setup VTA correctly (Yann)
```

```[tasklist]
### 2 Implement standard-based module mocking (Kasper)
- [x] Write public RFC to gain community feedback
- [x] Implement `storybook`, `stories` and `test` import/exports conditions in webpack and vite
- [x] **Demo** Make an module mocking example where the data-access module (the util, the service, the adapter, or whatever you call it) is mocked
- [x] **Demo** Make an module mocking example where all the DB calls are mocked with an in-memory DB (https://github.com/demonsters/prisma-mock)
- [x] **Test** Add module mocking to package.json imports in every sandbox, and have a simple story that uses that
- [ ] https://github.com/storybookjs/storybook/issues/26795
- [ ] https://github.com/storybookjs/storybook/issues/26796
- [ ] https://github.com/storybookjs/storybook/issues/10142
```

```[tasklist]
### 3 Add beforeEach as a new concept to CSF (Kasper)
- [x] Write public RFC to gain community feedback
- [x] Add CSF types: beforeRender: `(context: Context) => Awaitable<void>`
- [x] beforeRender as a function that runs after all loaders. First run preview, than meta, than story `beforeRender`
- [ ] https://github.com/storybookjs/storybook/issues/26723
- [x] **Test** Add unit tests for portable stories
- [x] **Nice to have** Implement the return value of beforeRender as a cleanup function, that is run when the component unmounts
- [x] **Demo** Experiment with `beforeEach` in the demo, and make a good example
- [x] **Test** Use beforeEach in a template story, and assert the execution is right, and that it is loaded after loaders and also test this in an e2e test, to check the order in the action panel (Yann)
- [ ] https://github.com/storybookjs/storybook/issues/26797
- [ ] https://github.com/storybookjs/storybook/issues/26994
```

-----
## Out of scope

```[tasklist]
### 4 Add a `mock` utility to storybook test that makes a module "mockable" (nice to have)
- [ ] Make a mock utility that wraps every function export in a spy
- [ ] Expand the mock utility to traverse object export so that every function on the objects becomes a spy as well (deeply)
- [ ] Give every spy we attach a mock name that references both the module name and the name of the function
- [ ] Make sure that every spy created with storybook/test is fully reactive, so that we can listen to calls and show them in the action panel.
- [ ] **Demo** Experiment with the `mock` utiltiy in the demo, and make a good example
- [ ] **Nice to have:** Deeply clone modules so that we can reset back all values, even primitive values before each story (similar as what vitest/jest do)
- [ ] **Docs** Add docs
- [ ] **Test** Add lots of unit tests for this utility
```

```[tasklist]
### 5 Make sure that $-prefixed-args are treated as non-component args in CSF (nice to have)
- [ ] Write public RFC to gain community feedback
- [ ] **Research** What kind of APIs are possible in Vue, should we make an API that the renderer can implement called `filterComponentArgs`
- [ ] Adjust csf types so that $-prefixed-args ares stripped from the first argument of the render function. Change the name of the args parameter to be called componentArgs
- [ ] Implement a default `filterComponentArgs` function for every renderer, that just strips out $-prefixed-args to be passed as first argument to the render function
- [ ] Override `filterComponentArgs` in Vue, so that componentArgs, and non-component args stay reactive.
- [ ] **Demo** Experiment with non-component args in the demo and make a good example
- [ ] **Docs** Add docs for non-component args.
- [ ] **Test** Use non component args in a template story, and assert that it is not passed to component, but is available to beforeRender and decerators
```

```[tasklist]
### 6 Implement sb mock command (nice to have)
- [ ] Plan this project out more, once we are sure we can scope it in
- [ ] `sb mock ./services/todos.ts` should autogenerate a mock file in `./services/todos.mock.ts` with all boilerplate using the mock utiltiy
- [ ] **Docs** Add docs
- [ ] **Test**
```
